include/tvm/runtime/container.h

```diff
@@ -0,0 +1,282 @@
+#include <utility>
+#include <vector>
+
+namespace tvm {
+namespace runtime {
+
+/*!
+ * \brief Base template for classes with array like memory layout.
+ *
+ *        It provides general methods to access the memory. The memory
+ *        layout is ArrayType + [ElemType]. The alignment of ArrayType
+ *        and ElemType is handled by the memory allocator.
+ *
+ * \tparam ArrayType The array header type, contains object specific metadata.
+ * \tparam ElemType The type of objects stored in the array right after
+ * ArrayType.
+ *
+ * \code
+ * // Example usage of the template to define a simple array wrapper
+ * class ArrayObj : public InplaceArrayBase<ArrayObj, Elem> {
+ * public:
+ *  // Wrap EmplaceInit to initialize the elements
+ *  template <typename Iterator>
+ *  void Init(Iterator begin, Iterator end) {
+ *   size_t num_elems = std::distance(begin, end);
+ *   auto it = begin;
+ *   for (size_t i = 0; i < num_elems; ++i) {
+ *     InplaceArrayBase::EmplaceInit(i, *it++);
+ *   }
+ *  }
+ * }
+ *
+ * void test_function() {
+ *   vector<Elem> fields;
+ *   auto ptr = make_inplace_array_object<ArrayObj, Elem>(fields.size());
+ *   ptr->Init(fields.begin(), fields.end());
+ *
+ *   // Access the 0th element in the array.
+ *   assert(ptr->operator[](0) == fields[0]);
+ * }
+ *
+ * \endcode
+ */
+template <typename ArrayType, typename ElemType>
+class InplaceArrayBase {
+ public:
+  /*!
+   * \brief Access element at index
+   * \param idx The index of the element.
+   * \return Const reference to ElemType at the index.
+   */
+  const ElemType& operator[](size_t idx) const {
+    size_t size = Self()->size();
+    CHECK_LT(idx, size) << "Index " << idx << " out of bounds " << size << "\n";
+    return *(reinterpret_cast<ElemType*>(AddressOf(idx)));
+  }
+
+  /*!
+   * \brief Access element at index
+   * \param idx The index of the element.
+   * \return Reference to ElemType at the index.
+   */
+  ElemType& operator[](size_t idx) {
+    size_t size = Self()->size();
+    CHECK_LT(idx, size) << "Index " << idx << " out of bounds " << size << "\n";
+    return *(reinterpret_cast<ElemType*>(AddressOf(idx)));
+  }
+
+  /*!
+   * \brief Destroy the Inplace Array Base object
+   */
+  ~InplaceArrayBase() {
+    if (!(std::is_standard_layout<ElemType>::value &&
+          std::is_trivial<ElemType>::value)) {
+      size_t size = Self()->size();
+      for (size_t i = 0; i < size; ++i) {
+        ElemType* fp = reinterpret_cast<ElemType*>(AddressOf(i));
+        fp->ElemType::~ElemType();
+      }
+    }
+  }
+
+ protected:
+  /*!
+   * \brief Construct a value in place with the arguments.
+   *
+   * \tparam Args Type parameters of the arguments.
+   * \param idx Index of the element.
+   * \param args Arguments to construct the new value.
+   */
+  template <typename... Args>
+  void EmplaceInit(size_t idx, Args&&... args) {
+    CHECK_LT(idx, Self()->capacity()) << "InplaceArray out of capacity\n";
+    void* field_ptr = AddressOf(idx);
```

Comment:
I start to think a bit more about EmplaceInit, I think it may not be the best function, mainly because of the exception safety problem.

Assume that we have already changed the size field to 1, and we are trying to init idx=0, then if the constructor throws an exception, the size remains at 1, while the space is not initialized. Then the destructor will incorrectly invoke the destruction of the corresponding element.

To make the code exception safe, we can introduce
```
template <typename... Args>
void EmplacePushBack(Args&& args) {
   // initialize the location size, then increase size by 1
   // the caller should check size is within capacity.
}
```

Reply:
The sub-class must always by default initialize size to 0

Reply:
I agree there is exception safety issue here. You solution is like my previous version, `InplaceArrayBase` calls childclass's `size` and `set_size` to manage the size, so under exception case, it will only destructs elements that are initialized. Another solution is to delegate the responsibility to child class, the child class increment the size one by one while calling `EmplaceInit`. The reason I prefer the second solution a bit is that `EmplacePushBack` is essentially `emplace_back`, which makes the template semantically look like a mutable object( I guess it's against the goal to put `push_back`, `assign`, `empalce_back` in child class). `EmplaceInit` is clearer that  it's only used for initialization, the object itself is still immutable.  What do you think?

Reply:
Also another interesting issue(bug) I found is that for ADT, we have to have capacity field in the object definition, otherwise it's impossible to have `size` and `capacity` at the same time while maintain exception safety. Right now we only have `size` field, it's used in both `size` and `capacity`. It's fine without exception. But under exception, we will try to destruct object that's not initialized anyway.

Reply:
OK, please ignore the second comment, you mentioned to check the capacity in child class.

Reply:
Yes, I think we can delegate the size management to the child class, however, please do add comment in the EmplaceInit function about this complication.